Libraries/LibGUI/IconView.cpp

```diff
@@ -210,17 +260,46 @@ bool IconView::update_rubber_banding(const Gfx::IntPoint& position)
 {
     auto adjusted_position = to_content_position(position);
     if (m_rubber_band_current != adjusted_position) {
+        auto prev_rect = Gfx::IntRect::from_two_points(m_rubber_band_origin, m_rubber_band_current);
         m_rubber_band_current = adjusted_position;
         auto rubber_band_rect = Gfx::IntRect::from_two_points(m_rubber_band_origin, m_rubber_band_current);
-        selection().clear();
-        for (auto item_index : items_intersecting_rect(rubber_band_rect)) {
-            selection().add(model()->index(item_index, model_column()));
-        }
-        if (m_rubber_banding_store_selection) {
-            for (auto stored_item : m_rubber_band_remembered_selection) {
-                selection().add(stored_item);
+
+        // Clearing the entire selection every time is very expensive,
+        // determine what items may need to be deselected and what new
+        // items may need to be selected. Avoid a ton of allocations.
+
+        auto deselect_area = prev_rect.shatter(rubber_band_rect);
+        auto select_area = rubber_band_rect.shatter(prev_rect);
+
+        // Initialize all candidate's toggle flag. We need to know which
+        // items we touched because the various rectangles likely will
+        // contain the same item more than once
+        for_each_item_intersecting_rects(deselect_area, [](ItemData& item_data) -> IterationDecision {
+            item_data.selection_toggled = false;
+            return IterationDecision::Continue;
+        });
+        for_each_item_intersecting_rects(select_area, [](ItemData& item_data) -> IterationDecision {
+            item_data.selection_toggled = false;
+            return IterationDecision::Continue;
+        });
+
+        // Now toggle all items that are no longer in the selected area, once only
+        for_each_item_intersecting_rects(deselect_area, [&](ItemData& item_data) -> IterationDecision {
+            if (!item_data.selection_toggled && item_data.is_intersecting(prev_rect) && !item_data.is_intersecting(rubber_band_rect)) {
+                item_data.selection_toggled = true;
+                toggle_selection(item_data);
             }
-        }
+            return IterationDecision::Continue;
+        });
+        // Now toggle all items that are in the new selected area, once only
+        for_each_item_intersecting_rects(select_area, [&](ItemData& item_data) -> IterationDecision {
+            if (!item_data.selection_toggled && !item_data.is_intersecting(prev_rect) && item_data.is_intersecting(rubber_band_rect)) {
+                item_data.selection_toggled = true;
+                toggle_selection(item_data);
```

Comment:
No, you want toggle_selection every time because when you add to the selection holding the Ctrl key, you want to be able to invert selected items as a way to deselect some

Reply:
Also, you need the `selection_toggled` flag because you only want to toggle each item once, because it's possible that you match the same item more than once during that loop. That's because the `Rect::hammer` may give you more than one rectangle that intersects with that item.